napari/util/keybindings.py

```diff
@@ -0,0 +1,582 @@
+PageUp, PageDown, Insert, Delete, Home, End, Escape, Backspace, F1,
+F2, F3, F4, F5, F6, F7, F8, F9, F10, F11, F12, Space, Enter, and Tab
+
+Functions take in only one argument: the parent that the function
+was bound to. This is the viewer or layer.
+
+By default, all functions are assumed to work on key presses only,
+but can be denoted to work on release too by separating the function
+into two statements with the yield keyword::
+
+    @viewer.bind_key('h')
+    def hello_world(viewer):
+        # on key press
+        viewer.status = 'hello world!'
+
+        yield
+
+        # on key release
+        viewer.status = 'goodbye world :('
+
+"""
+
+import re
+import types
+from collections import OrderedDict
+from enum import Enum
+from typing import Sequence, Iterable, Mapping, ByteString, Callable
+
+from vispy.util import keys
+
+from .misc import formatdoc
+
+
+STRINGTYPES = str, ByteString
+
+DOC_EXPLANATION = __doc__.replace('\n', '\n    ')
+
+
+SPECIAL_KEYS = [keys.SHIFT,
+                keys.CONTROL,
+                keys.ALT,
+                keys.META,
+                keys.UP,
+                keys.DOWN,
+                keys.LEFT,
+                keys.RIGHT,
+                keys.PAGEUP,
+                keys.PAGEDOWN,
+                keys.INSERT,
+                keys.DELETE,
+                keys.HOME,
+                keys.END,
+                keys.ESCAPE,
+                keys.BACKSPACE,
+                keys.F1,
+                keys.F2,
+                keys.F3,
+                keys.F4,
+                keys.F5,
+                keys.F6,
+                keys.F7,
+                keys.F8,
+                keys.F9,
+                keys.F10,
+                keys.F11,
+                keys.F12,
+                keys.SPACE,
+                keys.ENTER,
+                keys.TAB]
+
+MODIFIER_KEYS = OrderedDict(C=keys.CONTROL,
+                            M=keys.ALT,
+                            S=keys.SHIFT,
+                            s=keys.META)
+
+
+shorthand_modifier_patt = re.compile(f"([{''.join(MODIFIER_KEYS)}])(?=-)")
+
+
+class BindOperation(Enum):
+    BIND = 'bind'
+    UNBIND = 'unbind'
+    REBIND = 'rebind'
+
+
+def _expand_shorthand(shorthand_seq):
+    """Expand shorthand form of a key sequence.
+
+    C -> Control
+    M -> Alt
+    S -> Shift
+    s -> Meta
+    """
+    return re.sub(shorthand_modifier_patt,
+                  lambda m: MODIFIER_KEYS[m.group(0)].name,
+                  shorthand_seq)
+
+
+def parse_seq(seq):
+    """Parse a key sequence into its components in a comparable format.
+
+    Parameters
+    ----------
+    seq : str
+        Key sequence.
+
+    Returns
+    -------
+    key : str
+        Base key of the sequence.
+    modifiers : set of str
+        Modifier keys of the sequence.
+    """
+    parsed = re.split('-(?=.+)', seq)
+    modifiers, key = parsed[:-1], parsed[-1]
+
+    return key, set(modifiers)
+
+
+def components_to_seq(key, modifiers):
```

Comment:
a) some doctest examples would be super useful
b) as discussed in the call, let's rename sequence to combination

In some sense it's hard to distinguish between components and a key combination. My suggestion would be something like `normalize_key_combo`.

Reply:
LMAO I just saw that that's the next function... Maybe I should stop using "Add single comment" =P